[MFAC] training aware implementation for GradSampler

bfineran opened the pull request:
* updates the `GradSampler` class to allow for more flexible inputs to models with `args` and `kwargs`
* allows for `GradSampler`s to be passed on recipe initialization so that they may be run before pruning steps to avoid the need for gradient buffering during training. if no grad samplers are passed, normal gradient buffering will occur
* will need to do some small live tests, and then get an integration with our BERT flows

Commit:
[MFAC] training aware implementation for GradSampler

Commit:
quality

spacemanidol commented on research/mfac/tutorials/one_shot_pruning_with_mfac.md:
```diff
@@ -118,11 +131,12 @@ data_loader = DataLoader(dataset, shuffle=True, batch_size=16)
-def loss_function(x, y):
-    return torch.nn.functional.cross_entropy(x[0], y)
+def loss_function(model_outputs, loss_target):
+    return torch.nn.functional.cross_entropy(model_outputs[0], loss_target)
```
shouldn't we have an assert here that model_outputs is a tuple? Also worth adding a check that the shapes of loss target and model outputs match right?

spacemanidol reviewed (approved):
Looks good. Minor comment about adding an assert on outputs vs target outputs but feel free to ignore

Commit:
grad sampler reference fix

KSGulin committed:
M-FAC BERT Integration (#540)

Integrated MFAC Into BERT
Updated flow to use small block version in all reasonable cases
